### PR TITLE
v0.1.0/user api

### PR DESCRIPTION
**Pull request by duypv98**
Haven't implemented errorHandling and created new API spec yet

### CODE REVIEW

**Comment by duypv98** [comment]
App: username: duypv98, password: admin123456

File: controllers/events.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import eventCore from '../core/events';
```
Comment (namdaoduy) [inline controllers/events.js]: Please use absolute path (update all other places that using relative path)

File: core/encryption.js [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import bcrypt from 'bcrypt';
+
+const saltRounds = 12;
+
+function encrypt(plainText) {
+  const salt = bcrypt.genSaltSync(saltRounds);
+  const hash = bcrypt.hashSync(plainText, salt);
+  return hash;
```
Comment (namdaoduy) [inline core/encryption.js]: Why not return the salt? I think we must save the salt in DB

File: routes/events.js [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// import { Router } from 'express';
+// import eventController from '../controllers/events';
+
+// const router = Router();
```
Comment (namdaoduy) [inline routes/events.js]: Please update or remove

File: middlewares/verifyToken.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+function verifyToken(req, res, next) {
+  const bearerHeader = req.headers.authorization;
+  if (typeof bearerHeader !== 'undefined') {
```
Comment (namdaoduy) [inline middlewares/verifyToken.js]: `if (bearerHeader)`

And if you want compare something with `undefined`, just `(bearerHeader !== undefined)`

**Comment by namdaoduy** [comment]
Please update `core` and re-request review. I haven't review `core` yet because it's not finished

File: .babelrc [hunk 5]
```diff
@@ -2,7 +2,15 @@
...
     ["module-resolver", {
-      "root": ["./"]
+      "root": ["./"],
+      "alias": {
+        "db": "./db",
+        "core": "./core",
+        "middlewares": "./middlewares",
+        "routes": "./routes",
+        "controllers": "./controllers",
+        "utils": "./utils"
```
Comment (namdaoduy) [inline .babelrc]: Do we need alias when we already set `root` dir?

File: controllers/events.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import eventCore from 'core/events';
+
+const getEventsByUserId = async (req, res, next) => {
+  try {
+    const dataResponse = await eventCore.findEventsByUser(req.params.userId);
+    res.json(dataResponse);
```
Comment (namdaoduy) [inline controllers/events.js]:
```
res.json({
      data: dataResponse,
    });
```

File: core/users.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
+// import { CastError } from 'mongoose';
```
Comment (namdaoduy) [inline core/users.js]: Remove comment

File: core/users.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+import jwt from 'jsonwebtoken';
+import encryption from 'utils/encryption';
+import { UserNotFoundError, PasswordNotMatchError } from 'common/error';
+import User from 'db/models/User';
+
+const secretKey = process.env.SECRET_KEY;
```
Comment (namdaoduy) [inline core/users.js]: Put `SECRET_KEY` in config

File: core/users.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ *
+ * Find an user that matches given userId
+ * @param {String} userId
+ * @returns {Object} response data: User
+ */
+async function findById(userId) {
```
Comment (namdaoduy) [inline core/users.js]: `findUserById`

File: core/users.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    { _id: userId },
+    { _id: 0, password_hashed: 0, password_salt: 0 },
+  );
+
+  if (!user) {
+    const err = new UserNotFoundError();
+    throw err;
+  }
+  data = user;
+  return data;
```
Comment (namdaoduy) [inline core/users.js]: Unnecessary variable `data`

File: core/users.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+async function findById(userId) {
+  let data = {};
+  const user = await User.findOne(
+    { _id: userId },
+    { _id: 0, password_hashed: 0, password_salt: 0 },
+  );
+
+  if (!user) {
+    const err = new UserNotFoundError();
+    throw err;
```
Comment (namdaoduy) [inline core/users.js]: `throw new UserNotFoundError();` is enough

File: core/users.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  }
+  data = user;
+  return data;
+}
+
+async function checkLogin(usr, pwd) {
+  let data = {};
+  const user = await User.findOne({ username: usr });
+  if (!user) {
+    const err = new UserNotFoundError({ code: 40402, message: 'Invalid Username' });
```
Comment (namdaoduy) [inline core/users.js]: throw new Error, no need to create a variable

File: core/users.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    throw err;
+  }
+  if (encryption.isEqual(pwd, user.get('password_hashed'))) {
```
Comment (namdaoduy) [inline core/users.js]: `user.get('password_hashed')`? The model do not have method to get or set for each field?
We should not use string (constant) to get a property

File: middlewares/verifyToken.js [hunk 14]
```diff
@@ -0,0 +1,18 @@
+function verifyToken(req, res, next) {
+  const bearerHeader = req.headers.authorization;
+  if (bearerHeader !== 'undefined') {
```
Comment (namdaoduy) [inline middlewares/verifyToken.js]: `undefined` is not a string

Comment (namdaoduy) [reply middlewares/verifyToken.js]: `if (bearerHeader)` is enough

File: middlewares/verifyToken.js [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+    const token = bearerHeader.split(' ')[1];
```
Comment (namdaoduy) [inline middlewares/verifyToken.js]: Validate `bearerHeader.split(' ')` before get `[1]`. Can be a bug here if I provide an invalid token

File: middlewares/verifyToken.js [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+    req.token = token;
+    next();
```
Comment (namdaoduy) [inline middlewares/verifyToken.js]: I don not see any logic to verify token here? Just extract token from header?

File: routes/events.js (hunk 3 above)
Comment (namdaoduy) [reply routes/events.js]: Update before resolve.
Please add a comment and I will resolve it by myself if I think the updates are verified.

File: routes/users.js [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+const router = Router();
+
+router.route('/:userId')
+  .get(userController.getUserById);
+
+router.route('/:userId/events')
+  .get(eventController.getEventsByUserId);
+
+router.route('/login')
+  .post(userController.login);
```
Comment (namdaoduy) [inline routes/users.js]: Add a line break here

File: .babelrc (hunk 5 above)
Comment (duypv98) [reply .babelrc]: Okay, it's not neccessary. Removed

File: controllers/events.js (hunk 6 above)
Comment (duypv98) [reply controllers/events.js]: Fixed

File: core/users.js (hunk 7 above)
Comment (duypv98) [reply core/users.js]: Removed

File: core/users.js (hunk 8 above)
Comment (duypv98) [reply core/users.js]: Fixed

File: middlewares/verifyToken.js (hunk 14 above)
Comment (duypv98) [reply middlewares/verifyToken.js]: fixed

File: core/users.js (hunk 9 above)
Comment (duypv98) [reply core/users.js]: renamed

File: middlewares/verifyToken.js (hunk 15 above)
Comment (duypv98) [reply middlewares/verifyToken.js]: If a invalid token gave, how should I handle it ?

File: routes/users.js (hunk 17 above)
Comment (duypv98) [reply routes/users.js]: Fixed

File: middlewares/verifyToken.js (hunk 14 above)
Comment (duypv98) [reply middlewares/verifyToken.js]: Have not pused yet, as `verifyToken` logic has not been fixed

File: .env.example [hunk 18]
```diff
@@ -0,0 +1,7 @@
+DB_HOST=<Database Host>
+DB_PORT=<Database Port>
+DB_USER=<Database Username>
+DB_PWD=<Database Password>
+PORT=<NodeJs app. PORT>
+APP_ENV=<Application environtment: local, development, production>
+SECRET_KEY=<secretKey>
```
Comment (namdaoduy) [inline .env.example]: Change to JWT_SECRET_KEY

File: configs/dev.js [hunk 19]
```diff
@@ -1,12 +1,13 @@
...
-  DB_HOST, DB_PORT, DB_USER, DB_PWD,
+  DB_HOST, DB_PORT, DB_USER, DB_PWD, SECRET_KEY,
 } = process.env;
 
 const DB_URL = `mongodb://${DB_HOST}:${DB_PORT}/easy-event-v2`;
 const devConfig = {
   DB_URL,
   DB_USER,
   DB_PWD,
+  SECRET_KEY,
```
Comment (namdaoduy) [inline configs/dev.js]: Create a file `configs/base.js` that contain those config from `.env`
--> Now it's duplicated

File: core/users.js [hunk 20]
```diff
@@ -0,0 +1,51 @@
+import configs from 'configs/index';
+import jwt from 'jsonwebtoken';
+import encryption from 'utils/encryption';
+import { UserNotFoundError, PasswordNotMatchError } from 'common/error';
+import User from 'db/models/User';
+
+const secretKey = configs.SECRET_KEY;
```
Comment (namdaoduy) [inline core/users.js]: No need to use a variable. Using `configs.SECRET_KEY` where you need it is enough

File: middlewares/verifyToken.js [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+function verifyToken(req, res, next) {
+  const bearerHeader = req.headers.authorization;
+  if (bearerHeader) {
+    const splitedHeader = bearerHeader.split(' ');
+    const token = splitedHeader[1];
+    jwt.verify(token, secretKey, (err, result) => {
+      if (err) {
+        if (err instanceof TokenExpiredError) {
+          throw new InvalidTokenError({ code: 40302, message: 'TokenExpired' });
```
Comment (namdaoduy) [inline middlewares/verifyToken.js]: Use 2 different type of error class for 2 error

File: .env.example (hunk 18 above)
Comment (duypv98) [reply .env.example]: Fixed

File: configs/dev.js (hunk 19 above)
Comment (duypv98) [reply configs/dev.js]: Please check my new update and review again

File: core/users.js (hunk 20 above)
Comment (duypv98) [reply core/users.js]: Fixed

File: middlewares/verifyToken.js (hunk 21 above)
Comment (duypv98) [reply middlewares/verifyToken.js]: Done

Comment (duypv98) [reply middlewares/verifyToken.js]: https://github.com/EasyEventV2/ee2_api/pull/1/commits/2f849054ef50719c8e970815dd43b0004ed521d4

File: core/users.js (hunk 20 above)
Comment (duypv98) [reply core/users.js]: https://github.com/EasyEventV2/ee2_api/pull/1/commits/73716262368593e9f1d6f9ac94ab472189fd237b

File: configs/dev.js (hunk 19 above)
Comment (duypv98) [reply configs/dev.js]: https://github.com/EasyEventV2/ee2_api/pull/1/commits/211a357d9c7cbf2f9d23f130d288f01545cdd4af

File: .env.example (hunk 18 above)
Comment (duypv98) [reply .env.example]: https://github.com/EasyEventV2/ee2_api/commit/73716262368593e9f1d6f9ac94ab472189fd237b

File: configs/dev.js [hunk 22]
```diff
@@ -1,12 +1,9 @@
...
-} = process.env;
+import baseConfig from 'configs/base';
 
-const DB_URL = `mongodb://${DB_HOST}:${DB_PORT}/easy-event-v2`;
+const DB_URL = `mongodb://${baseConfig.DB_HOST}:${baseConfig.DB_PORT}/easy-event-v2`;
 const devConfig = {
   DB_URL,
-  DB_USER,
-  DB_PWD,
+  ...baseConfig,
```
[the remark this answers is not in the document]
Comment (duypv98) [reply configs/dev.js]: fixed in https://github.com/EasyEventV2/ee2_api/pull/1/commits/b426d88c9fe424ffcb851fd87cd8ab40dffa5254

File: core/users.js (hunk 11 above)
Comment (duypv98) [reply core/users.js]: fixed in https://github.com/EasyEventV2/ee2_api/pull/1/commits/94f76d867f0e4b41358e9c942e9fc3c9053d102b

File: core/users.js (hunk 10 above)
Comment (duypv98) [reply core/users.js]: fixed in https://github.com/EasyEventV2/ee2_api/pull/1/commits/94f76d867f0e4b41358e9c942e9fc3c9053d102b

File: core/users.js (hunk 12 above)
Comment (duypv98) [reply core/users.js]: fixed https://github.com/EasyEventV2/ee2_api/pull/1/commits/94f76d867f0e4b41358e9c942e9fc3c9053d102b

**Pull request merged**